src/OrbitGl/ScopeTree.cpp

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ScopeTree.h"
+
+#include "OrbitBase/Tracing.h"
+
+using orbit_client_protos::TimerInfo;
+
+ScopeTree::ScopeTree() {
+  static TimerInfo kDefaultTimerInfo;
+  root_ = CreateNode(&kDefaultTimerInfo);
+}
+
+ScopeNode* ScopeTree::CreateNode(const TimerInfo* timer_info) {
+  nodes_.push_back(ScopeNode(timer_info, size_++));
+  ScopeNode* node = nodes_.Last();
+  return node;
+}
+
+void ScopeTree::Insert(const TimerInfo& timer_info) { root_->Insert(CreateNode(&timer_info)); }
+
+void ScopeTree::Print() const {
+  LOG("Printing %u nodes depth=%u:", size_, Depth());
+  root_->Print();
+  LOG("");
+}
+
+void ScopeNode::Print(const ScopeNode* node, uint64_t start_time, uint32_t depth) {
+  LOG("%s%lu ScopeNode(%u) %p [%lu, %lu]", std::string(depth, ' '), start_time, node->id_, node,
+      node->timer_info_->start(), node->timer_info_->end());
+  for (auto& pair : node->children_) {
+    Print(pair.second, pair.first, depth + 1);
+  }
+}
+
+void ScopeNode::FindDepth(const ScopeNode* node, size_t* max_depth, size_t current_depth) {
```

Comment:
O(n), consider saving depth in the nodes if you want this operation to be faster